File: Sebastián_V/Sebastian_Python/Stack y queue.py
# =========================================
# 🔹 STACK (PILA)
# =========================================

class Stack:
    def __init__(self):
        self.data = []

    def push(self, item):
        self.data.append(item)

    def pop(self):
        if self.is_empty():
            print("El stack está vacío")
            return None
        return self.data.pop()

    def peek(self):
        if self.is_empty():
            print("El stack está vacío")
            return None
        return self.data[-1]

    def is_empty(self):
        return len(self.data) == 0

    def size(self):
        return len(self.data)

    def show(self):
        print(self.data)


# =========================================
# 🔹 QUEUE (COLA)
# =========================================

class Queue:
    def __init__(self):
        self.data = []

    def enqueue(self, item):
        self.data.append(item)

    def dequeue(self):
        if self.is_empty():
            print("La queue está vacía")
            return None
        return self.data.pop(0)  # elimina el primero

    def front(self):
        if self.is_empty():
            print("La queue está vacía")
            return None
        return self.data[0]

    def is_empty(self):
        return len(self.data) == 0

    def size(self):
        return len(self.data)

    def show(self):
        print(self.data)


# =========================================
# 🔹 PRUEBA
# =========================================

if __name__ == "__main__":
    print("=== STACK ===")
    s = Stack()
    s.push(10)
    s.push("hola")
    s.push(3.14)
    s.show()

    print("Pop:", s.pop())
    print("Top:", s.peek())
    s.show()

    print("\n=== QUEUE ===")
    q = Queue()
    q.enqueue(1)
    q.enqueue("mundo")
    q.enqueue(2.5)
    q.show()

    print("Dequeue:", q.dequeue())
    print("Front:", q.front())
    q.show()

File: Sebastián_V/Sebas-Rust/src/Stack_y_queue.rs
// =========================================
// 🔹 STACK (PILA)
// =========================================

struct Stack {
    data: Vec<String>,
}

impl Stack {
    fn new() -> Self {
        Stack { data: Vec::new() }
    }

    fn push(&mut self, item: String) {
        self.data.push(item);
    }

    fn pop(&mut self) -> Option<String> {
        self.data.pop()
    }

    fn peek(&self) -> Option<&String> {
        self.data.last()
    }

    fn is_empty(&self) -> bool {
        self.data.is_empty()
    }

    fn size(&self) -> usize {
        self.data.len()
    }

    fn show(&self) {
        println!("{:?}", self.data);
    }
}


// =========================================
// 🔹 QUEUE (COLA)
// =========================================

struct Queue {
    data: Vec<String>,
}

impl Queue {
    fn new() -> Self {
        Queue { data: Vec::new() }
    }

    fn enqueue(&mut self, item: String) {
        self.data.push(item);
    }

    fn dequeue(&mut self) -> Option<String> {
        if self.is_empty() {
            None
        } else {
            Some(self.data.remove(0)) // elimina el primero
        }
    }

    fn front(&self) -> Option<&String> {
        self.data.first()
    }

    fn is_empty(&self) -> bool {
        self.data.is_empty()
    }

    fn size(&self) -> usize {
        self.data.len()
    }

    fn show(&self) {
        println!("{:?}", self.data);
    }
}


// =========================================
// 🔹 PRUEBA
// =========================================

pub(crate) fn stack() {
    println!("=== STACK ===");
    let mut s = Stack::new();

    s.push("10".to_string());
    s.push("hola".to_string());
    s.push("3.14".to_string());

    s.show();

    println!("Pop: {:?}", s.pop());
    println!("Top: {:?}", s.peek());

    s.show();

    println!("\n=== QUEUE ===");
    let mut q = Queue::new();

    q.enqueue("1".to_string());
    q.enqueue("mundo".to_string());
    q.enqueue("2.5".to_string());

    q.show();

    println!("Dequeue: {:?}", q.dequeue());
    println!("Front: {:?}", q.front());

    q.show();
}